sdk/azidentity/client_secret_credential.go

```diff
@@ -54,48 +54,65 @@ func (c *ClientSecretCredential) AuthenticationPolicy(options azcore.Authenticat
 }
 
 type bearerTokenPolicy struct {
-	// take lock when manipulating header/expiresOn fields
-	lock      sync.RWMutex
-	header    string
-	expiresOn time.Time
+	empty     chan bool    // indicates we don't have a token yet
+	updating  atomic.Int64 // atomically set to 1 to indicate a refresh is in progress
+	header    atomic.String
+	expiresOn atomic.Time
 	creds     azcore.TokenCredential // R/O
 	scopes    []string               // R/O
 }
 
 func newBearerTokenPolicy(creds azcore.TokenCredential, scopes []string) *bearerTokenPolicy {
-	// set the token as expired so first call to Do() refreshes it
-	return &bearerTokenPolicy{creds: creds, scopes: scopes, expiresOn: time.Now().UTC()}
+	bt := bearerTokenPolicy{creds: creds, scopes: scopes}
+	// prime channel indicating there is no token
+	bt.empty = make(chan bool, 1)
+	bt.empty <- true
+	return &bt
 }
 
 func (b *bearerTokenPolicy) Do(ctx context.Context, req *azcore.Request) (*azcore.Response, error) {
-	var bt string
-	// take read lock and check if the token has expired
-	b.lock.RLock()
+	// check if the token is empty
+	if <-b.empty {
+		tk, err := b.creds.GetToken(ctx, b.scopes)
+		if err != nil {
+			// failed to get a token, let other go routines try
+			b.empty <- true
+			return nil, err
+		}
+		b.header.Store("Bearer " + tk.Token)
+		b.expiresOn.Store(tk.ExpiresOn)
+		// signal token has been initialized.  go routines
+		// that read from b.empty will always get false.
+		close(b.empty)
+	}
+	// create a "refresh window" before the token's real expiration date.
+	// this allows callers to continue to use the old token while the
+	// refresh is in progress.
+	const window = 2 * time.Minute
+	// check if the token has expired
 	now := time.Now().UTC()
-	if now.Equal(b.expiresOn) || now.After(b.expiresOn) {
-		// token has expired, take the write lock then check again
-		b.lock.RUnlock()
-		// don't defer Unlock(), we want to release it ASAP
-		b.lock.Lock()
-		if now.Equal(b.expiresOn) || now.After(b.expiresOn) {
-			// token has expired, get a new one and update shared state
+	exp := b.expiresOn.Load()
+	if now.After(exp.Add(-window)) {
```

Comment:
you could do this:
time.Now().UTC().After(exp.Add(-window))

And get rid of line 93